File: app/src/main/java/khorunzhyicom/alex/theposter/service/common/InjectableCommandActionService.java
package khorunzhyicom.alex.theposter.service.common;


import io.techery.janet.ActionHolder;
import io.techery.janet.ActionService;
import io.techery.janet.ActionServiceWrapper;
import io.techery.janet.JanetException;
import khorunzhyicom.alex.theposter.application.App;
import khorunzhyicom.alex.theposter.di.components.ApiComponent;
import khorunzhyicom.alex.theposter.di.components.injector.CommandInjector;
import khorunzhyicom.alex.theposter.service.commands.InjectableCommand;

public class InjectableCommandActionService extends ActionServiceWrapper {

    private final Injector injector;

    public InjectableCommandActionService(ActionService service, App app) {
        super(service);
        this.injector = Injector.create((app.apiComponent()));
    }

    @Override
    protected <A> boolean onInterceptSend(ActionHolder<A> holder) {
        A action = holder.action();
        if (!(action instanceof InjectableCommand)) return false;
        injector.inject((InjectableCommand) action);
        return false;
    }

    @Override
    protected <A> void onInterceptCancel(ActionHolder<A> holder) {}

    @Override
    protected <A> void onInterceptStart(ActionHolder<A> holder) {}

    @Override
    protected <A> void onInterceptProgress(ActionHolder<A> holder, int progress) {}

    @Override
    protected <A> void onInterceptSuccess(ActionHolder<A> holder) {}

    @Override
    protected <A> boolean onInterceptFail(ActionHolder<A> holder, JanetException e) {
        return false;
    }

    private static class Injector {

        private final CommandInjector injector;

        private Injector(CommandInjector injector) {
            this.injector = injector;
        }

        public static Injector create(ApiComponent apiComponent) {
            return new Injector(apiComponent);
        }

        public void inject(InjectableCommand command) {
            command.injectTo(injector);
        }
    }
}

File: app/src/main/java/khorunzhyicom/alex/theposter/service/common/InjectableCommandActionService.kt
package khorunzhyicom.alex.theposter.service.common


import io.techery.janet.ActionHolder
import io.techery.janet.ActionService
import io.techery.janet.ActionServiceWrapper
import io.techery.janet.JanetException
import khorunzhyicom.alex.theposter.App
import khorunzhyicom.alex.theposter.di.components.ApiComponent
import khorunzhyicom.alex.theposter.di.components.injector.CommandInjector
import khorunzhyicom.alex.theposter.service.commands.base.InjectableCommand

class InjectableCommandActionService(service: ActionService, app: App) : ActionServiceWrapper(service) {

    private val injector: Injector

    init {
        this.injector = Injector.create(app.apiComponent())
    }

    override fun <A> onInterceptSend(holder: ActionHolder<A>): Boolean {
        val action = holder.action() as? InjectableCommand ?: return false
        injector.inject(action)
        return false
    }

    override fun <A> onInterceptCancel(holder: ActionHolder<A>) {}

    override fun <A> onInterceptStart(holder: ActionHolder<A>) {}

    override fun <A> onInterceptProgress(holder: ActionHolder<A>, progress: Int) {}

    override fun <A> onInterceptSuccess(holder: ActionHolder<A>) {}

    override fun <A> onInterceptFail(holder: ActionHolder<A>, e: JanetException): Boolean {
        return false
    }

    private class Injector private constructor(private val injector: CommandInjector) {

        fun inject(command: InjectableCommand) = command.injectTo(injector)

        companion object {

            fun create(apiComponent: ApiComponent): Injector = Injector(apiComponent)

        }
    }
}
